### PR TITLE
Fix #1332 - Add autoinsert support

### LINKED ISSUE #1332: [VUE] Add autoinsert support
Volar have built'in support for tag, brace and other insertion (volar/client/autoInsert), similar to HTML LS I have a plan to add support for this

### PR DESCRIPTION
**Pull request by zulus**
I started from AutoInsert reconciler, but as the result , there is a quiet big path:

1. AutoInsert support
2. Original VSCode preferences mapper (plus registered preferences in Eclipse) - I'm reading descriptions and types from vscode plugin package.json and map to IEclipsePreferences
3. Vue server can consume preferences from CSS/HTML/JS/TS and probably others (see fallbackCollect)
4. Special map for "typescript-server" settings to vscode settings aliases (vue require names as in vscode settings.json)
5. typescript-server-vue plugin

In general, preference support is nighmare. If server can consume extra preferences, each section server have to be mapped manually. I think this should be somehow pluggable, probably on LSP4E level

### CODE REVIEW

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/vue/autoinsert/VueAutoInsertReconciler.java
Comment (angelozerr) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/vue/autoinsert/VueAutoInsertReconciler.java:74]: This code that I de for XML is so bad and can give some blocking thread bugs.

@mickaelistria LSP4e should provide an API to execute somerhing after à didChange

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/vue/autoinsert/VueAutoInsertReconciler.java
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/vue/autoinsert/VueAutoInsertReconciler.java:104]: Can be
```suggestion
															String newText = response.map(Function.identity(), r::getNewText);
```

**Review by mickaelistria — APPROVED** [review]
I don't know much about vue; but it looks OK to me (I've just put 1 comment about code but it's not important).
So I think it's OK to merge it if build succeeds.

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/vue/autoinsert/VueAutoInsertReconciler.java
Comment (zulus) [reply org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/vue/autoinsert/VueAutoInsertReconciler.java:74]: I off course agree, for now this is copy-paste from HTMLAutoInsertReconciler

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/vue/autoinsert/VueAutoInsertReconciler.java
Comment (zulus) [reply org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/vue/autoinsert/VueAutoInsertReconciler.java]: Done, as I wrote in description this patch is much bigger than AutoInsert. Take a look to VuePreferences, and how it work with vscode-vue package.json

Comment (mickaelistria) [reply org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/vue/autoinsert/VueAutoInsertReconciler.java]: > Done, as I wrote in description this patch is much bigger than AutoInsert. 

Do you think it's worth reworking this patch in a series of easier to understand yet functionally valuable patches to facilitate further maintenance?

Comment (zulus) [reply org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/vue/autoinsert/VueAutoInsertReconciler.java]: Sure, one moment

Comment (zulus) [reply org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/vue/autoinsert/VueAutoInsertReconciler.java]: I published autoinsert-only part (works on default settings). For example auto close html tags `<tag>|</tag>`, and transform `{{|}}` to `{{ | }}`  in templates

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/vue/autoinsert/VueAutoInsertReconciler.java
Comment (zulus) [reply org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/vue/autoinsert/VueAutoInsertReconciler.java:75]: I made some tests and prepared pull request for lsp4e: https://github.com/eclipse/lsp4e/pull/812

**Comment by mickaelistria** [comment]
OK, feel free to merge this one as soon as you're ready for it ;)

**Comment by mickaelistria** [comment]
> Sure, when I will gain privileges :D

OK, I see the nomination process pretty advanced but not complete yet (requires EMO or PMC to give a final approval I guess). Do you prefer I merge it now, or you merge it later?

**Comment by zulus** [comment]
Please merge as-is. THX

**Pull request merged**
